### ISSUE TITLE
Tornado AssertionError when starting salt-minion on deb/wheezy (develop)

### ISSUE THREAD

**Issue by bechtoldt** [saltstack/salt#24440]
I'm trying to the test my changes from #24422 using the latest develop branch.

```
./bootstrap-salt.sh -P -M -K -g https://github.com/bechtoldt/salt.git git refactor_es_modules
```

```
=> master1: Downloading/unpacking tornado>=4.0
==> master1:   Running setup.py egg_info for package tornado
==> master1:
==> master1:     no previously-included directories found matching 'docs/build'
==> master1:     warning: no files found matching 'tornado/test/README'
==> master1: Downloading/unpacking backports.ssl-match-hostname (from tornado>=4.0)
==> master1:   Downloading backports.ssl_match_hostname-3.4.0.2.tar.gz
==> master1:   Running setup.py egg_info for package backports.ssl-match-hostname
==> master1:
==> master1: Downloading/unpacking certifi (from tornado>=4.0)
==> master1:   Running setup.py egg_info for package certifi
==> master1:
==> master1: Installing collected packages: tornado, backports.ssl-match-hostname, certifi
==> master1:   Running setup.py install for tornado
==> master1:     building 'tornado.speedups' extension
==> master1:     gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -I/usr/include/python2.7 -c tornado/speedups.c -o build/temp.linux-x86_64-2.7/tornado/speedups.o
==> master1:     tornado/speedups.c:49:1: warning: function declaration isn’t a prototype [-Wstrict-prototypes]
==> master1:     gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-z,relro build/temp.linux-x86_64-2.7/tornado/speedups.o -o build/lib.linux-x86_64-2.7/tornado/speedups.so
==> master1:
==> master1:     no previously-included directories found matching 'docs/build'
==> master1:     warning: no files found matching 'tornado/test/README'
==> master1:   Running setup.py install for backports.ssl-match-hostname
==> master1:
==> master1:   Running setup.py install for certifi
==> master1:
==> master1: Successfully installed tornado backports.ssl-match-hostname certifi
```

```
# salt-minion --versions
Salt Version:
           Salt: 2015.5.2-5695-g2c2cc20

Dependency Versions:
         Jinja2: 2.6
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.10
          PyZMQ: 13.1.0
         Python: 2.7.3 (default, Mar 13 2014, 11:03:55)
           RAET: Not Installed
            ZMQ: 3.2.3
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.1.10
       pycrypto: 2.6

System Versions:
           dist: debian 7.8
        machine: x86_64
        release: 3.2.0-4-amd64
         system: debian 7.8
```

```

# pip freeze
Warning: cannot find svn location for distribute==0.6.24dev-r0
Jinja2==2.6
M2Crypto==0.21.1
MarkupSafe==0.15
PyYAML==3.10
argparse==1.2.1
backports.ssl-match-hostname==3.4.0.2
certifi==2015.04.28
chardet==2.0.1
## FIXME: could not find svn URL in dependency_links for this package:
distribute==0.6.24dev-r0
msgpack-python==0.1.10
pycrypto==2.6
python-apt==0.8.8.2
pyzmq==13.1.0
requests==2.0.0
salt==2015.5.2-5695-g2c2cc20
six==1.1.0
tornado==4.2
urllib3==1.7.1
wsgiref==0.1.2
```

```
2015-06-05 11:15:14,072 [salt.scripts                             ][ERROR   ][25853] Minion failed to start:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 81, in minion_process
    minion.start()
  File "/usr/lib/python2.7/dist-packages/salt/cli/daemons.py", line 269, in start
    self.minion.tune_in()
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1593, in tune_in
    self.sync_connect_master()
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 589, in sync_connect_master
    raise six.reraise(*future_exception)
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 879, in run
    yielded = self.gen.send(value)
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 597, in connect_master
    yield self._post_master_init(master)
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 242, in wrapper
    Runner(result, future, yielded)
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 817, in __init__
    if self.handle_yield(first_yielded):
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 956, in handle_yield
    self.future = convert_yielded(yielded)
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 1022, in convert_yielded
    return multi_future(yielded)
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 643, in multi_future
    assert all(is_future(i) for i in children)
AssertionError
2015-06-05 11:15:14,073 [salt.scripts                             ][WARNING ][25853] ** Restarting minion **
```

**Comment by rallytime** [comment]
Thanks for this report @bechtoldt. Are you able to install a newer version of PyZMQ? I do no know for sure, but I know that sometimes the 14.x versions clear up some of these errors.

@jacksontj @thatch45 @cachedout - Have we decided on which version of PyZMQ we will be depending on in Beryllium? I think the requirements files say 13.1, but I think there are still some bugs to work out with that version if that is the case. There are still some ZMQ/transport-related tests failing on Jenkins on Develop (namely Debian 7 and Ubuntu 12, which both run PyZMQ 13.x, and not 14, by default.)

**Comment by jacksontj** [comment]
Personally I'd like to require PyZMQ 14.x+, and last I heard there were discussions amongst you guys what we wanted to support. The plan (IIRC) is to make zmq an optional dep by the next feature release.

**Comment by bechtoldt** [comment]
same problem on centos7?

```
# salt-minion --versions
Salt Version:
           Salt: 2015.5.2-6416-g58ffb6d

Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.3.1
         Python: 2.7.5 (default, Jun 17 2014, 18:11:42)
           RAET: Not Installed
        Tornado: 4.1
            ZMQ: 3.2.5
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
       pycrypto: 2.6.1

System Versions:
           dist: centos 7.1.1503 Core
        machine: x86_64
        release: 3.10.0-229.1.2.el7.x86_64
         system: CentOS Linux 7.1.1503 Core
```

```
2015-06-19 19:04:05,357 [salt.scripts     ][WARNING ][1988] ** Restarting minion **
2015-06-19 19:04:06,757 [salt.scripts     ][ERROR   ][2049] Minion failed to start:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/scripts.py", line 81, in minion_process
    minion.start()
  File "/usr/lib/python2.7/site-packages/salt/cli/daemons.py", line 269, in start
    self.minion.tune_in()
  File "/usr/lib/python2.7/site-packages/salt/minion.py", line 1597, in tune_in
    self.sync_connect_master()
  File "/usr/lib/python2.7/site-packages/salt/minion.py", line 589, in sync_connect_master
    raise six.reraise(*future_exception)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 812, in run
    yielded = self.gen.send(value)
  File "/usr/lib/python2.7/site-packages/salt/minion.py", line 597, in connect_master
    yield self._post_master_init(master)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 224, in wrapper
    Runner(result, future, yielded)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 756, in __init__
    if self.handle_yield(first_yielded):
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 886, in handle_yield
    self.future = convert_yielded(yielded)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 952, in convert_yielded
    return multi_future(yielded)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 596, in multi_future
    assert all(is_future(i) for i in children)
AssertionError
2015-06-19 19:04:06,758 [salt.scripts     ][WARNING ][2049] ** Restarting minion **
2015-06-19 19:04:09,157 [salt.scripts     ][ERROR   ][2110] Minion failed to start:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/scripts.py", line 81, in minion_process
    minion.start()
  File "/usr/lib/python2.7/site-packages/salt/cli/daemons.py", line 269, in start
    self.minion.tune_in()
  File "/usr/lib/python2.7/site-packages/salt/minion.py", line 1597, in tune_in
    self.sync_connect_master()
  File "/usr/lib/python2.7/site-packages/salt/minion.py", line 589, in sync_connect_master
    raise six.reraise(*future_exception)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 812, in run
    yielded = self.gen.send(value)
  File "/usr/lib/python2.7/site-packages/salt/minion.py", line 597, in connect_master
    yield self._post_master_init(master)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 224, in wrapper
    Runner(result, future, yielded)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 756, in __init__
    if self.handle_yield(first_yielded):
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 886, in handle_yield
    self.future = convert_yielded(yielded)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 952, in convert_yielded
    return multi_future(yielded)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 596, in multi_future
    assert all(is_future(i) for i in children)
AssertionError
2015-06-19 19:04:09,157 [salt.scripts     ][WARNING ][2110] ** Restarting minion **
...
```

**Comment by jacksontj** [comment]
@bechtoldt What does your minion config look like (mostly caring about the `master`, `transport`, and `pub_port` sections.

**Comment by bechtoldt** [comment]
@jacksontj 

```
# cat /etc/salt/master
autosign_file: /etc/salt/autosign.conf

file_roots:
  base:
    - /srv/salt/states
    - /srv/salt/contrib/states
    - /vagrant/shared

pillar_roots:
  base:
    - /srv/salt/pillar/examples
    - /srv/salt/pillar/shared
[root@client1 ~]# cat /etc/salt/minion
file_roots:
  base:
    - /srv/salt/states
    - /srv/salt/contrib/states
    - /vagrant/shared

pillar_roots:
  base:
    - /srv/salt/pillar/examples
    - /srv/salt/pillar/shared

module_dirs:
  - /srv/salt/_modules/common
  - /srv/salt/_modules/formulas

states_dirs:
  - /srv/salt/_states/common

file_client: local
```

**Comment by jacksontj** [comment]
Hmm, this is going to take more digging than I have time this morning... basically the issue seems to be that ``yield self._post_master_init(master)`` is attempting to yield a non-future (meaning _post_master_init() is returning not a future). I'm not sure how that happens, nor am I able to reproduce easily-- so I'll have to circle back to this after I think about it some more (or figure out a way to repro)

**Comment by bechtoldt** [comment]
@jacksontj Maybe I'll find some time to create a vagrant image (vbox) that can be used to reproduce this issue.

**Comment by jacksontj** [comment]
That would be immensely helpful :)

**Comment by bechtoldt** [comment]
@jacksontj https://github.com/bechtoldt/salt-debug-24440, hope this helps!

**Comment by jacksontj** [comment]
We need to get more bug reports with vagrant setups -- that was very helpful! Fix is in #25148, basically the switch to async pillars broke local file_client minions.

**Comment by bechtoldt** [comment]
:+1:
